pkg/scheduler/api/hyper_node_info.go

```diff
@@ -321,14 +334,90 @@ func (hni *HyperNodesInfo) getChildren(hyperNodeName string) sets.Set[string] {
 	return children
 }
 
+// GetLeafNodes returns the leaf hyperNodes set for a given hyperNode.
+func (hni *HyperNodesInfo) GetLeafNodes(hyperNodeName string) sets.Set[string] {
+	leafNodes := sets.New[string]()
+	ancestorsChain := sets.New[string]()
+	hni.findLeafNodesWithCycleCheck(hyperNodeName, leafNodes, ancestorsChain)
+	return leafNodes
+}
+
+func (hni *HyperNodesInfo) findLeafNodesWithCycleCheck(hyperNodeName string, leafNodes sets.Set[string], ancestorsChain sets.Set[string]) {
+	if ancestorsChain.Has(hyperNodeName) {
+		klog.ErrorS(nil, "Cycle detected in HyperNode hierarchy", "hyperNode", hyperNodeName)
+		return
+	}
+
+	hn, ok := hni.hyperNodes[hyperNodeName]
+	if !ok {
+		return
+	}
+
+	ancestorsChain.Insert(hyperNodeName)
+	defer ancestorsChain.Delete(hyperNodeName)
+
+	isLeaf := true
+	for _, member := range hn.HyperNode.Spec.Members {
+		if member.Type == topologyv1alpha1.MemberTypeHyperNode {
+			isLeaf = false
+			hni.findLeafNodesWithCycleCheck(member.Selector.ExactMatch.Name, leafNodes, ancestorsChain)
+		}
+	}
+
+	if isLeaf {
+		leafNodes.Insert(hyperNodeName)
+	}
+}
+
+// GetRegexSelectorLeafHyperNodes returns leaf hyperNodes whose member's selector is regex match.
+func (hni *HyperNodesInfo) GetRegexSelectorLeafHyperNodes() sets.Set[string] {
+	leaf := sets.New[string]()
+	for name, hnInfo := range hni.hyperNodes {
+		if hnInfo == nil || hnInfo.HyperNode == nil {
+			continue
+		}
+
+		isLeaf := true
+		hasRegexMatch := false
+		for _, member := range hnInfo.HyperNode.Spec.Members {
+			if member.Type == topologyv1alpha1.MemberTypeHyperNode {
+				isLeaf = false
+				break
+			}
+			if member.Selector.RegexMatch != nil {
+				hasRegexMatch = true
+			}
+		}
+		if isLeaf && hasRegexMatch {
+			leaf.Insert(name)
+		}
+	}
+	return leaf
+}
+
 // setParent sets the parent of a HyperNode member.
-func (hni *HyperNodesInfo) setParent(hn *HyperNodeInfo, memberName string) error {
-	parent, ok := hni.parentMap[memberName]
-	if ok && parent != hn.Name {
-		return fmt.Errorf("HyperNode %s already has a parent %s, and cannot set another parent %s", memberName, parent, hn.Name)
-	} else {
-		hni.parentMap[memberName] = hn.Name
+func (hni *HyperNodesInfo) setParent(member, parent string) error {
+	hn, ok := hni.hyperNodes[member]
+	if !ok {
+		klog.InfoS("HyperNode not exists in cache, maybe not created or not be watched, will set parent first", "name", member, "parent", parent)
+		hn = NewHyperNodeInfo(&topologyv1alpha1.HyperNode{ObjectMeta: metav1.ObjectMeta{
+			Name: member,
+		}})
+		hni.hyperNodes[member] = hn
+		hn.parent = parent
+		return nil
 	}
+	currentParent := hn.parent
+	if currentParent == "" {
+		hn.parent = parent
+		return nil
+	}
+
+	if currentParent != parent {
+		hni.builtErrHyperNode = parent
```

Comment:
What if the later HyperNode is the correct one, and the `builtErrHyperNode` is set to this Hypernode that is actually the correct Hypernode. Then in `updateAncestors` `if hni.builtErrHyperNode == name` won't be triggered if I only reconfigured the wrong HyperNode

Reply:
I think we can infomer user that they should restart volcano scheduler when set multi parent, and fixed it in next version.

Reply:
ok